### PR TITLE
refactor(dcutr): remove `direct::Handler`

### LINKED ISSUE #4013: fix(dcutr): remove direct handler
## Description

Remove `libp2p_dcutr::handler::direct`.

## Motivation

The only reason why we have `libp2p_dcutr::handler::direct::Handler` is to emit an event once the direct connection is established. This has been needed, but with recent refactorings in `libp2p-swarm` one might as well just emit the event from the `libp2p_dcutr::Behaviour` right away.

See https://github.com/libp2p/rust-libp2p/pull/3982#discussion_r1210090137 for details.

## Are you planning to do it yourself in a pull request?

No

### PR DESCRIPTION
**Pull request by tcoratger**
## Description

With recent refactorings in `libp2p-swarm`, there is no more need for `handler::direct` at all.

Resolves #4013.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/dcutr/CHANGELOG.md [hunk 1]
```diff
@@ -4,8 +4,11 @@
   See [PR 3715].
 - Remove deprecated items. See [PR 3700].
 
+- Remove `handler::direct`. See [PR 4016].
```
Comment (thomaseizinger) [inline protocols/dcutr/CHANGELOG.md]: Not sure if we need an entry for this? I'd consider that an implementation detail :)

File: protocols/dcutr/src/behaviour_impl.rs [hunk 2]
```diff
@@ -383,21 +376,7 @@ impl NetworkBehaviour for Behaviour {
...
-                        peer_id: event_source,
-                        handler: NotifyHandler::One(relayed_connection_id),
-                        event: Either::Left(
-                            handler::relayed::Command::UpgradeFinishedDontKeepAlive,
-                        ),
-                    },
-                    ToSwarm::GenerateEvent(Event::DirectConnectionUpgradeSucceeded {
-                        remote_peer_id: event_source,
-                    }),
-                ]);
```
Comment (thomaseizinger) [inline protocols/dcutr/src/behaviour_impl.rs:398]: This snippet will need to happen where we previously instantiated `direct::Handler`.

**Review by thomaseizinger — DISMISSED** [review]
Thank you! Two comments :)

File: protocols/dcutr/src/behaviour_impl.rs (hunk 2 above)
Comment (tcoratger) [reply protocols/dcutr/src/behaviour_impl.rs:398]: @thomaseizinger Yes right, I removed that as never reached but in fact, now, `Either::Right` can be either unreachable or corresponding to where we previously instantiated `direct::Handler`.

File: protocols/dcutr/CHANGELOG.md (hunk 1 above)
Comment (tcoratger) [reply protocols/dcutr/CHANGELOG.md]: @thomaseizinger I hesitated but as it concerns indirectly public stuffs, I pushed a changelog. But no problem I've just removed it.

File: protocols/dcutr/src/behaviour_impl.rs (hunk 2 above)
Comment (thomaseizinger) [reply protocols/dcutr/src/behaviour_impl.rs:398]: Not quite right unfortunately.

We want to emit these events when we established a direction connection. Previously, that was only possible with a dedicated `ConnectionHandler` implementation. Now we can just push these events into `queued_events` inside `handle_established_outbound_connection` where we previously said `return direct::Handler` and now return `dummy::ConnectionHandler`.

If you look at the implementation of `direct::Handler`, it doesn't actually do anything other than emitting an event which is used here again to emit one to the behaviour.

We can just not do this back and forth and simply push these events into `queued_events` directly.

Comment (tcoratger) [reply protocols/dcutr/src/behaviour_impl.rs:398]: @thomaseizinger Ok I think I understand (I hope) now. Before, we used the `on_connection_handler_event` to do a match to know what to push as event into `queued_events`. But now, we don't need that anymore and so we can directly push the event into `queued_events` when we establish the outbound connection while returning `dummy::ConnectionHandler`, it doesn't matter now that the we pushed the event correctly.

Sorry for the misunderstandings, the code is super clear but there are complicated parts and I lack experience on it so a lot of questions must seem stupid to you

**Review by mxinden — APPROVED** [review]
Very cool to see the fruits of the past `libp2p-swarm` refactorings. Thank you @tcoratger for picking this up!

File: protocols/dcutr/src/behaviour_impl.rs (hunk 2 above)
Comment (mxinden) [reply protocols/dcutr/src/behaviour_impl.rs:398]: > the code is super clear

As the author of most of `libp2p-dcutr`, I don't think my code is very clear. Especially the indirection that you just removed in this pull request is very confusing.

> a lot of questions must seem stupid to you

Not at all. I am impressed by how well you understand the code-base @tcoratger especially given the short period of time you have been involved in the project.

**Pull request merged**